ambry-store/src/main/java/com.github.ambry.store/PersistentIndex.java

```diff
@@ -1056,6 +1140,45 @@ void persistIndex() throws StoreException {
     persistor.write();
   }
 
+  /**
+   * Gets the list of {@link IndexSegment} files that refer to the log segment with name {@code logSegmentName}.
+   * @param dataDir the directory where the index files are.
+   * @param logSegmentName the name of the log segment whose index segment files are required.
+   * @return the list of {@link IndexSegment} files that refer to the log segment with name {@code logSegmentName}.
+   */
+  static File[] getIndexSegmentFilesForLogSegment(String dataDir, final String logSegmentName) {
+    return new File(dataDir).listFiles(new FilenameFilter() {
+      @Override
+      public boolean accept(File dir, String name) {
+        return name.startsWith(logSegmentName) && name.endsWith(IndexSegment.INDEX_SEGMENT_FILE_NAME_SUFFIX);
+      }
+    });
+  }
+
+  /**
+   * Cleans up all files related to index segments that refer to the log segment with name {@code logSegmentName}.
+   *  @param dataDir the directory where the index files are.
+   * @param logSegmentName the name of the log segment whose index segment related files need to be deleteds.
+   * @throws IOException if {@code dataDir} could not be read or if a file could not be deleted.
+   */
+  static void cleanupIndexSegmentFilesForLogSegment(String dataDir, final String logSegmentName) throws IOException {
+    File[] filesToCleanup = new File(dataDir).listFiles(new FilenameFilter() {
+      @Override
+      public boolean accept(File dir, String name) {
+        return name.startsWith(logSegmentName) && (name.endsWith(IndexSegment.INDEX_SEGMENT_FILE_NAME_SUFFIX)
+            || name.endsWith(IndexSegment.BLOOM_FILE_NAME_SUFFIX));
+      }
+    });
+    if (filesToCleanup == null) {
+      throw new IOException("Failed to list index segment files");
+    }
+    for (File file : filesToCleanup) {
+      if (!file.delete()) {
+        throw new IOException("Could not delete file named " + file);
```

Comment:
That is for the caller to handle - the store is an inconsistent state and will require inspection. I don't think there is any reason to keep deleting the other files since a bad state has been reached.